File: DataStructures/Trees/BFS・DFS/atcoder/A62/A62.php
<?php
fscanf(STDIN, "%d %d", $N, $M);

// 隣接リストを初期化
$graph = array_fill(1, $N, []);
$visited = array_fill(1, $N, false);

// 辺情報を読み込む
for ($i = 0; $i < $M; $i++) {
    fscanf(STDIN, "%d %d", $a, $b);
    $graph[$a][] = $b;
    $graph[$b][] = $a; // 無向グラフなので逆方向も追加
}

// DFS定義（再帰）
function dfs($v, &$graph, &$visited)
{
    $visited[$v] = true;
    foreach ($graph[$v] as $neighbor) {
        if (!$visited[$neighbor]) {
            dfs($neighbor, $graph, $visited);
        }
    }
}

// 頂点1からDFS開始
dfs(1, $graph, $visited);

// 連結判定
if (in_array(false, $visited, true)) {
    echo "The graph is not connected.\n";
} else {
    echo "The graph is connected.\n";
}

File: DataStructures/Trees/BFS・DFS/atcoder/A62/A62.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	graph   [][]int
	visited []bool
)

func dfs(v int) {
	visited[v] = true
	for _, neighbor := range graph[v] {
		if !visited[neighbor] {
			dfs(neighbor)
		}
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	parts := strings.Fields(scanner.Text())
	N, _ := strconv.Atoi(parts[0])
	M, _ := strconv.Atoi(parts[1])

	// 初期化
	graph = make([][]int, N+1)
	visited = make([]bool, N+1)

	// 辺の読み取り
	for i := 0; i < M; i++ {
		scanner.Scan()
		parts = strings.Fields(scanner.Text())
		a, _ := strconv.Atoi(parts[0])
		b, _ := strconv.Atoi(parts[1])
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a) // 無向グラフなので両方向
	}

	// DFSで連結成分を探索（頂点1から）
	dfs(1)

	// 判定
	allVisited := true
	for i := 1; i <= N; i++ {
		if !visited[i] {
			allVisited = false
			break
		}
	}

	if allVisited {
		fmt.Println("The graph is connected.")
	} else {
		fmt.Println("The graph is not connected.")
	}
}
